electron: Add libnotify, pipewire, and several other runtime dependencies to RPATH; fixes desktop notifications and screenshare

mmarx opened the pull request:
mattermost-desktop is missing the libnotify runtime dependency, so desktop notifications are not shown.

## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

teutat3s commented:
@ofborg build mattermost-desktop

SuperSandro2000 commented on pkgs/applications/networking/instant-messengers/mattermost-desktop/default.nix:
```diff
@@ -53,6 +60,7 @@ stdenv.mkDerivation {
       --replace /share/mattermost-desktop/mattermost-desktop /bin/mattermost-desktop
 
     makeWrapper '${lib.getExe electron_28}' $out/bin/${pname} \
+      --set LD_LIBRARY_PATH "${libPath}" \
```
It's there an alternative to usong LD_LIBRARY_PATH?

teutat3s commented:
I did not look into alternatives yet, but this approach was used for element-desktop as well: https://github.com/NixOS/nixpkgs/pull/301522/files

mmarx commented:
In looking for alterternatives, I found that ultimately, `libnotify` is loaded via `dlopen(3)`. I guess the proper fix would be to just have it in `electron-unwrapped`'s `RPATH`, which likely requires some manual `patchelf` invocation. I can look into that, but it will take a few days until I have access to a machine I can reasonably build `electron` on.

SuperSandro2000 commented:
Can we try that? Using patchelf is highly preferred over LD_LIBRARY_PATH.

teutat3s commented:
@mmarx if you have an idea and a diff to share, I'd have a build machine available to try it.

teutat3s commented:
@SuperSandro2000 Your comment made me think if there could be a `RPATH` + `patchelf` solution for #303725 as well?

mmarx commented:
> @SuperSandro2000 Your comment made me think if there could be a `RPATH` + `patchelf` solution in `electron` for #303725 as well?

yeah, that should work as well, it's also `dlopen`ing pipewire by default. I'm still traveling, but should be able to come up with a patch tonight.

mmarx committed:
electron: add libnotify to the RPATH

Electron loads libnotify dynamically via dlopen(3), which fails if it
is not in the RPATH. This, in turn, breaks desktop notifications in,
e.g., mattermost-desktop.

mmarx committed:
element-desktop: do not set LD_LIBRARY_PATH

libnotify is now included in electron's RPATH, so we can avoid setting
LD_LIBRARY_PATH here.

mmarx committed:
electron: add pipewire to RPATH

mmarx commented:
It's probably a good idea to also check the packages depending on `electron` for such `LD_LIBRARY_PATH` workarounds; I'll look into that tomorrow.

teutat3s commented:
Non exhaustive list of package I found from the root of this repo with this command:
```
grep -R --files-with-matches electron pkgs | xargs grep LD_LIBRARY_PATH
```

- `vesktop`
- `tidal-hifi`
- `webcord`
- `jitsi-meet-electron`
- `teams-for-linux`
- `signal-desktop`

Would this `rpath` fix only work for packages using nixpkgs source-built `electron` or also with packages that ship with it's own bundled `electron` version, like `signal-desktop`?

mmarx commented:
> Non exhaustive list of package I found from the root of this repo with this command:
> 
> ```
> grep -R --files-with-matches electron pkgs | xargs grep LD_LIBRARY_PATH
> ```
> 
>     * `vesktop`
> 
>     * `tidal-hifi`
> 
>     * `webcord`
> 
>     * `jitsi-meet-electron`
> 
>     * `teams-for-linux`
> 
>     * `signal-desktop`
> 

thanks, that's a good starting point.

> Would this `rpath` fix only work for packages using nixpkgs source-built `electron` or also with packages that ship with it's own bundled `electron` version, like `signal-desktop`?

I don't see a reason why it wouldn't also work with bundled `electron` versions, but of course those would need to have their own `patchelf` invocations.

teutat3s commented:
It looks like some packages also add `libpulseaudio` to `LD_LIBRARY_PATH` ...
- `teams-for-linux`
- `webcord`
- `vesktop`

teutat3s commented:
@mmarx I built `element-desktop` and `electron` with this PR's branch checked out and notifications and screen sharing works in my test. :+1:

teutat3s commented on pkgs/development/tools/electron/common.nix:
```diff
@@ -190,6 +190,19 @@ in (chromium.override { upstream-info = info.chromium; }).mkDerivation (base: {
     runHook postInstall
   '';
 
+  postFixup =
```
This wipes out chromium's `postFixup` phase, which we are overriding here.

teutat3s commented on pkgs/development/tools/electron/common.nix:
```diff
@@ -190,6 +190,19 @@ in (chromium.override { upstream-info = info.chromium; }).mkDerivation (base: {
+        pipewire
+      ];
+    in
+  ''
```
```suggestion
  base.postFixup + ''
```

mmarx committed:
jitsi-meet-electron: do not set LD_LIBRARY_PATH

mmarx committed:
electron: add libstdc++ to RPATH

mmarx committed:
electron: add libsecret to RPATH

mmarx committed:
electron: add libpulseaudio to RPATH

mmarx committed:
electron: add speechd to RPATH

mmarx committed:
webcord: do not set LD_LIBRARY_PATH

teutat3s commented:
Result of `nixpkgs-review pr 303590` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>8 packages built:</summary>
  <ul>
    <li>electron</li>
    <li>element-desktop</li>
    <li>jitsi-meet-electron</li>
    <li>standardnotes</li>
    <li>super-productivity</li>
    <li>teams-for-linux</li>
    <li>vesktop</li>
    <li>webcord</li>
  </ul>
</details>

SuperSandro2000 reviewed (approved):
Looks to me like a better solution than adding those to many electron applications.

mmarx commented on pkgs/development/tools/electron/common.nix:
```diff
@@ -190,6 +193,23 @@ in (chromium.override { upstream-info = info.chromium; }).mkDerivation (base: {
+        pipewire
+        stdenv.cc.cc.lib
+        libsecret
+        libpulseaudio
```
`teams-for-linux` and `webcord` had both in their `LD_LIBRARY_PATHS`; according to a comment, `teams` uses pulseaudio for meeting sound and pipewire for screensharing.
